### PR TITLE
Add wget replacement go-downloader

### PR DESCRIPTION
**Pull request by dmcilvaney**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [ ] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [ ] All package sources are available
- [ ] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [ ] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [ ] All source files have up-to-date hashes in the `*.signatures.json` files
- [ ] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [ ] Documentation has been updated to match any changes to the build system
- [ ] Ready to merge

---

###### Summary 

When gathering toolchain .rpms we use `wget`. While `wget` is generally quite resilient there are a few error conditions it can't recover from easily. We should retry the download if possible. A simple retry wrapper in bash would likely be sufficient but with an eye to the future (enabling caching etc.) we can just wrap the download operation in a simple go tool. This tool accepts a subset of arguments used by wget.

###### Change Log  

- Add `downloader` tool

###### Does this affect the toolchain?  

**YES** - Just downloading it with REBUILD_TOOLCHAIN=n

###### Associated issues  

- https://microsoft.visualstudio.com/OS/_workitems/edit/47288886

###### Test Methodology

- Local

### CODE REVIEW

**Review by anphel31 — APPROVED** [review]

**Review by trungams — APPROVED** [review]

**Review by neha170 — APPROVED** [review]
Not blocking, but having a test file would be nice for the new tool :)

File: toolkit/tools/downloader/downloader.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+	if *dstFile == "" {
+		// strip query strings from url
+		u, err := url.Parse(*srcUrl)
+		if err != nil {
+			logger.Log.Fatalf("Invalid URL (%s), error: %s", *srcUrl, err)
+		}
+		*dstFile = filepath.Base(u.Path)
+		if *prefixDir != "" {
+			*dstFile = filepath.Join(*prefixDir, *dstFile)
+		}
```
Comment (neha170) [inline toolkit/tools/downloader/downloader.go:97]: Both conditions are the same.

File: toolkit/tools/downloader/downloader.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+		downloadRetryAttempts = 6
+		failureBackoffBase    = 3.0
+		downloadRetryDuration = time.Second
+	)
+	var noCancel chan struct{} = nil
+
+	_, err = retry.RunWithExpBackoff(func() error {
+		netErr := network.DownloadFile(srcUrl, dstFile, caCerts, tlsCerts)
+		if netErr != nil {
+			logger.Log.Warnf("Failed to download (%s). Error: %s", srcUrl, netErr)
```
Comment (neha170) [inline toolkit/tools/downloader/downloader.go:121]: nit: having same style of error logging throughout the file would be nice.
'. Error:'   v/s    ', error:'     v/s     '. Error:\n'

File: toolkit/tools/downloader/downloader.go (hunk 1 above)
Comment (dmcilvaney) [reply toolkit/tools/downloader/downloader.go:97]: If you mean `*frefixDir != ""`, its intentional. I'll refactor slightly so it makes more sense.

File: toolkit/tools/downloader/downloader.go (hunk 2 above)
Comment (dmcilvaney) [reply toolkit/tools/downloader/downloader.go:122]: I'd like to keep these intentionally as one line, (Also, the logger can't use %w).
I can add newlines to the fatals though.

File: toolkit/tools/downloader/downloader.go (hunk 1 above)
Comment (neha170) [reply toolkit/tools/downloader/downloader.go:100]: Thanks for the refactor !

**Comment by dmcilvaney** [comment]
image
Added some warnings when we hit 404 errors so it doesn't look like the tool is stuck

**Pull request merged**